### ISSUE TITLE
miniterm.py fails on Windows Subsystem for Linux due to TIOCSTI 

### ISSUE THREAD

**Issue by thomasbrandon** [pyserial/pyserial#243]
On [Windows Subsystem for Linux](https://msdn.microsoft.com/en-us/commandline/wsl/about) the TIOCSTI ioctl isn't supported so Console.cancel fails like:
```
File "/usr/local/lib/python2.7/dist-packages/serial/tools/miniterm.py", line 165, in cancel
  fcntl.ioctl(self.fd, termios.TIOCSTI, b'\0')
IOError: [Errno 25] Inappropriate ioctl for device
```
I submitted https://github.com/Microsoft/BashOnWindows/issues/2332. But upon investigation this doesn't seem like a great ioctl to use given security issues with it and a general lack of use (googling it mainly gives articles saying it shouldn't be in linux). So not sure adding it to WSL is the best solution and an alternate solution within pyserial would seem better if possible but not sure how best to do that.
This ioctl was introduced in https://github.com/pyserial/pyserial/commit/cab3dabc81ecbf010d81ff38433db99c1f6d667c to replace the previous select based solution.
Happy to hack at it but pointers would be helpful.

**Comment by zsquareplusc** [comment]
That is not easy to replace... I removed the solution that used `select` because it interfered with the buffering in the io modules buffered text handler.

A way around may be to treat `stdin` as binary input, so `select` would work, and then push that into the text decoder, looping until a character pops out. extra steps are needed to find out the correct encoding. basically what miniterm is doing with the serial port... so much more code than the one liner with `TIOCSTI` ;-)

**Comment by thomasbrandon** [comment]
Is the IO buffering issue you refer to the fact that if you say paste to the terminal then select will only trigger one call to read(1) so all but the first character remain buffered?
I found that this seems to solve that without too much complication:
``` python
stdin = os.fdopen(self.fd, 'rb', 0)
self.enc_stdin = codecs.getreader(sys.stdin.encoding)(stdin)
```
(Where that codec stuff was already conditionally used for python 2)
Or am I missing more subtle issues (so far only tested outside of miniterm so I can trigger cancel).
Then I used an epoll rather than select with a flag set in cancel() and checked in getkey() to eliminate the pipe so while not as simple as the ioctl it is at least a little cleaner than the pipe solution.

**Comment by zsquareplusc** [comment]
it was in conjunction with escape sequences, e.g. from cursor keys. see #174

i also looked at `poll` (w/o `e`) but that is not implemented in at least one popular posix system. `epoll` seems to be only supported by linux.

**Comment by thomasbrandon** [comment]
Thanks for the pointer, will investigate further.

OK, re epoll, could also use select with a timeout to similarly avoid the need for the pipe.

**Comment by thomasbrandon** [comment]
Yeah, https://github.com/pyserial/pyserial/issues/174 also seems to be fixed by disabling buffering. With buffering the first getkey only returns the first character of the multiple character escape sequence while the next call (after more input triggers select again) returns the other characters (but again only the first character of the new sequence).
It looks to be fixed by changing ([here](https://github.com/pyserial/pyserial/blob/bfaf6c8b354beee34010a9be187756c60bfd09f9/serial/tools/miniterm.py#L149))
``` python
        if sys.version_info < (3, 0):
            self.enc_stdin = codecs.getreader(sys.stdin.encoding)(sys.stdin)
        else:
            self.enc_stdin = sys.stdin
```
to:
``` python
        stdin = os.fdopen(self.fd, 'rb', 0)
        self.enc_stdin = codecs.getreader(sys.stdin.encoding)(stdin)
```
But as the current implementation is otherwise cleaner, and miniterm.py hangs on WSL due to https://github.com/Microsoft/BashOnWindows/issues/2330 which looks like it needs MS fixes, I can't see any point in changing this now. If MS fix the other issue but not TIOSCTI then maybe.
Plus the tool I am looking at using ([idf-monitor.py](https://github.com/espressif/esp-idf/blob/master/tools/idf_monitor.py)) actually makes little use of miniterm apart from `Console.cancel()` and even that it re-implements if older miniterm versions are used so better to fix there (it also doesn't hang due to different threading logic). And in fact `cancel` only seems to be used in miniterm given errors, the other tool adds commands which use `cancel` to pause miniterm so fail to run due to this issue.

Thanks for the help @zsquareplusc

**Comment by zsquareplusc** [comment]
re timeout, miniterm used to use that on the serial port to check the alive flag from time to time. but that gives a CPU load >0.0% even if it is doing nothing. i don't like programs that use CPU to do nothing ;-)

**Comment by tavip** [comment]
TIOCSTI is now disabled by default on Linux as well (kernels >= 6.2)
